forage-android/src/main/java/com/joinforage/forage/android/ecom/services/vault/bt/BTResponseParser.kt

```diff
@@ -15,48 +18,61 @@ internal class BTResponseParser(btRes: Result<Any?>) : VaultResponseParser {
 
     val isSuccessful: Boolean
 
+    private val failureResponseRegExp = BtFailureResponseRegExp(btRes)
+
     init {
         // BT returns a Result<Any?> so it's never null
         vaultError = parseVaultError(btRes)
-        forageError = parseForageError(BaseResponseRegExp(btRes))
+        forageError = parseForageError(failureResponseRegExp)
         isSuccessful = btRes.isSuccess && vaultError == null && forageError == null
         successfulResponse = parseSuccessfulResponse(btRes)
     }
 
     private fun parseVaultError(vaultResponse: BasisTheoryResponse): ForageApiResponse.Failure? {
         if (vaultResponse.isSuccess) return null
-        val resRegExp = BtResponseRegExp(vaultResponse)
-        return try {
-            // In AbstractVaultSubmitter we track the forageError
-            // and the forage status code of the UnknownErrorApiResponse
-            // via the VaultProxyResponseMonitor. This keeps us informed
-            // of *when* erroroneous BT response happen. Unfortunately,
-            // we do not currently track the specifics of the proxy_error
-            // that that BT returned to us.
-            // TODO: log the specific error that BT responds with when
-            //  resRegExp.containsProxyError == true
-            if (resRegExp.containsProxyError) UnknownErrorApiResponse else null
-        } catch (_: Exception) {
-            null
-        }
+        // In AbstractVaultSubmitter we track the forageError
+        // and the forage status code of the UnknownErrorApiResponse
+        // via the VaultProxyResponseMonitor. This keeps us informed
+        // of *when* erroroneous BT response happen. Unfortunately,
+        // we do not currently track the specifics of the proxy_error
+        // that that BT returned to us.
+        // TODO: log the specific error that BT responds with when
+        //  resRegExp.containsProxyError == true
+        return if (failureResponseRegExp.containsProxyError) UnknownErrorApiResponse else null
     }
 
-    private fun parseForageError(resRegExp: BaseResponseRegExp): ForageApiResponse.Failure? {
-        if (resRegExp.bodyText == null || resRegExp.statusCode == null) return null
-        return try {
-            ForageApiResponse.Failure(resRegExp.statusCode, resRegExp.bodyText)
-        } catch (_: Exception) {
+    private fun parseForageError(resRegExp: BtFailureResponseRegExp): ForageApiResponse.Failure? =
+        if (resRegExp.bodyText == null ||
+            resRegExp.statusCode == null ||
+            // if there's a proxy_error, don't try to parse as a Forage error
+            resRegExp.containsProxyError
```

Comment:
I replace the try/catch with an additional clause in the if-statement that confirms `resRegExp.containsProxyError == false` before attempting to parse as a ForageError. It is true that attempting to parse an error that is not a Forage error with `ForageApiResponse.Failure(..)` will throw an exception but, if the error is not a proxy_error, then it can only be a ForageError and if we are unable to parse the FroageError, that is an exceptional circumstance and we need to throw.

NOTE: the callee of `BTResponseParser` will catch the exception, log it, and gracefully return `UnknownServerError`